### PR TITLE
Andy/mysql auto increment

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: enginetest/insert_queries.go [hunk 1]
```diff
@@ -441,6 +458,186 @@ var InsertScripts = []ScriptTest{
...
+		Name: "auto increment on tinyint unsigned",
+		SetUpScript: []string{
+			"create table auto (pk tinyint unsigned primary key auto_increment)",
+			"insert into auto values (NULL),(10),(0)",
+		},
+		Assertions: []ScriptTestAssertion{
+			{
+				Query: "select * from auto",
+				Expected: []sql.Row{
+					{uint64(1)},{uint64(10)},{uint64(11)},
```
Comment (zachmu) [inline enginetest/insert_queries.go:546]: Heads up that the current test implementation for row comparison widens all numeric types before comparison, so these casts are all no-ops. If you think this is important behavior to test you will have to come up with another way.

File: sql/expression/autoincrement.go [hunk 2]
```diff
@@ -8,23 +8,40 @@ import (
...
 }
 
 // NewAutoIncrement creates a new AutoIncrement expression.
-func NewAutoIncrement(lastInsertId, given sql.Expression) (*AutoIncrement, error) {
-	_, ok := lastInsertId.Type().(sql.NumberType)
+func NewAutoIncrement(ctx *sql.Context, table sql.Table, given sql.Expression) (*AutoIncrement, error) {
+	autoTbl, ok := table.(sql.AutoIncrementTable)
 	if !ok {
-		return nil, errors.New("AutoIncrement must be given a number type expression")
+		return nil, errors.New("this table does not support AUTO_INCREMENT columns")
```
Comment (zachmu) [inline sql/expression/autoincrement.go:24]: Error type plz

File: sql/expression/autoincrement.go [hunk 3]
```diff
@@ -8,23 +8,40 @@ import (
...
+		return nil, err
+	}
+
+	var autoCol *sql.Column
+	for _, c := range autoTbl.Schema() {
+		if c.AutoIncrement {
+			autoCol = c
+			break
+		}
 	}
```
Comment (zachmu) [inline sql/expression/autoincrement.go:38]: Should have an error check for a table that doesn't have any auto increment columns declared

File: sql/plan/show_indexes.go [hunk 4]
```diff
@@ -64,7 +64,10 @@ func (n *ShowIndexes) Schema() sql.Schema {
 func (n *ShowIndexes) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
 	table, ok := n.Child.(*ResolvedTable)
 	if !ok {
-		panic(fmt.Sprintf("unexpected type %T", n.Child))
+		table, ok = n.Child.Children()[0].(*ResolvedTable)
```
Comment (zachmu) [inline sql/plan/show_indexes.go:67]: This is awfully fragile. Check out getResolvedTable in the analyzer package (and maybe move it to the plan package while you're at it)

**Review by zachmu — APPROVED** [review]
LG, just a few comments

File: sql/plan/show_indexes.go (hunk 4 above)
Comment (andrew-wm-arthur) [reply sql/plan/show_indexes.go:67]: this was a workaround for the exchange node problems, it wasn't supposed to land here

**Pull request merged**